lib/dialects/sqlite3/schema/ddl.js

```diff
@@ -477,14 +442,47 @@ class SQLite3_DDL {
    * It'll be helpful to refactor this file heavily to combine/optimize some of these calls
    */
 
-  reinsertMapped(createTable, newSql, mapRow) {
-    return Promise.resolve()
-      .then(() => this.createTempTable(createTable))
-      .then(() => this.copyData())
-      .then(() => this.dropOriginal())
-      .then(() => this.trx.raw(newSql))
-      .then(() => this.reinsertData(mapRow))
-      .then(() => this.dropTempTable());
+  async reinsertMapped(createTable, newSql, mapRow) {
+    const wasForeignKeysEnabled =
+      (
+        await this.client.raw('PRAGMA foreign_keys').connection(this.connection)
+      )[0].foreign_keys === 1;
+
+    if (wasForeignKeysEnabled) {
+      await this.client
+        .raw('PRAGMA foreign_keys = OFF')
+        .connection(this.connection);
+    }
+
+    try {
+      await this.client.transaction(
+        async (trx) => {
+          await this.createTempTable(trx, createTable);
+          await this.copyData(trx);
+          await this.dropOriginal(trx);
+          await trx.raw(newSql);
+          await this.reinsertData(trx, mapRow);
+          await this.dropTempTable(trx);
+
+          if (wasForeignKeysEnabled) {
+            const foreignKeysViolations = await trx.raw(
+              'PRAGMA foreign_key_check'
+            );
+
+            if (foreignKeysViolations.length > 0) {
+              throw new Error('FOREIGN KEY constraint failed');
+            }
+          }
+        },
+        { connection: this.connection }
+      );
+    } finally {
```

Comment:
nice touch executing this in `finally` block.

Reply:
Just to be on the safe side! Unfortunately, this does not work with generated SQL statements.